File: backend/app/services/geo.py
"""Python port of src/lib/geo.js — geofence + distance math, plus the
synthetic-telemetry helper used by both the demo log generator and the
"location omitted" fallback in the usage-log endpoint."""

import math
from typing import Optional

EARTH_RADIUS_KM = 6371


def _to_rad(deg: float) -> float:
    return deg * math.pi / 180


def distance_km(a: Optional[dict], b: Optional[dict]) -> Optional[float]:
    if not a or not b:
        return None
    d_lat = _to_rad(b["lat"] - a["lat"])
    d_lng = _to_rad(b["lng"] - a["lng"])
    h = (
        math.sin(d_lat / 2) ** 2
        + math.cos(_to_rad(a["lat"])) * math.cos(_to_rad(b["lat"])) * math.sin(d_lng / 2) ** 2
    )
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(h))


def geofence_check(location: Optional[dict], site_id: Optional[str], site_by_id: dict) -> Optional[dict]:
    site = site_by_id.get(site_id) if site_id else None
    if not location or not site or location.get("lat") is None or location.get("lng") is None:
        return None

    km = distance_km(location, {"lat": site["lat"], "lng": site["lng"]})
    return {
        "site": site,
        "distanceKm": km,
        "radiusKm": site["radius_km"],
        "breach": km > site["radius_km"],
        "overshootKm": max(0.0, km - site["radius_km"]),
    }


def point_near_site(site_id: Optional[str], site_by_id: dict, offset_km: float = 0) -> Optional[dict]:
    site = site_by_id.get(site_id) if site_id else None
    if not site:
        return None
    return {
        "lat": round(site["lat"] + offset_km / 111, 5),
        "lng": site["lng"],
    }

File: src/lib/geo.js
import { siteById } from '../data/sites.js'

const EARTH_RADIUS_KM = 6371
const toRad = (deg) => (deg * Math.PI) / 180

/** Great-circle distance in km between two lat/lng points. */
export function distanceKm(a, b) {
  if (!a || !b) return null
  const dLat = toRad(b.lat - a.lat)
  const dLng = toRad(b.lng - a.lng)
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.sin(dLng / 2) ** 2
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(h))
}

/**
 * Test a logged position against the boundary of the site the unit is
 * assigned to. This is what turns "geofence breach" from a hardcoded
 * incident into something derived from the usage log.
 */
export function geofenceCheck(location, siteId) {
  const site = siteId ? siteById[siteId] : null
  if (!location || !site || location.lat == null || location.lng == null) return null

  const km = distanceKm(location, { lat: site.lat, lng: site.lng })
  return {
    site,
    distanceKm: km,
    radiusKm: site.radiusKm,
    breach: km > site.radiusKm,
    overshootKm: Math.max(0, km - site.radiusKm),
  }
}

/** A point offset from a site centre — used to simulate on-site / off-site pings. */
export function pointNearSite(siteId, offsetKm = 0) {
  const site = siteById[siteId]
  if (!site) return null
  // ~111 km per degree of latitude; offset due north keeps the maths obvious.
  return {
    lat: +(site.lat + offsetKm / 111).toFixed(5),
    lng: site.lng,
  }
}
